Feature/liayoo/account importing

liayoo opened the pull request:
- Support importing accounts from keystore files at `<BLOCKCHAIN_DATA_DIR>/keys/<PORT>`
- Update scripts for deploying & starting nodes on gcp
- [Design doc](https://docs.google.com/document/d/1Bad90noJiN4LOdwxwiW2eu0awANBpU6LjJ9aL_XNdgM/edit#)

platfowner commented on common/file-util.js:
```diff
@@ -200,6 +209,49 @@ class FileUtil {
+      logger.debug(`Failed to read keys dir (${keysPath}): ${err.stack}`);
+      return null;
+    }
+    files = files.filter((filename) => moment(filename, KEYSTORE_FILENAME_FORMAT).isValid());
```
Could you add some note about this filename format and filtering with it?

liayoo replied:
As discussed offline, this filtering is now deprecated. Thanks!

platfowner commented on node/index.js:
```diff
@@ -37,32 +39,64 @@ const Transaction = require('../tx-pool/transaction');
+      const accountFromKeystore = FileUtil.getLatestKeystoreAccount(this.keysDir, password);
+      if (accountFromKeystore !== null) {
+        this.account = accountFromKeystore;
+      } else {
```
suggestion: could we skip this account generation logic unless we have specific use cases? (this might make it difficult to notice some mistakes in deployment like forgetting to copy keystore files?)

liayoo replied:
Changed to expect either ACCOUNT_INDEX or KEYSTORE_FILE_PATH and not generate an account. Thanks!

platfowner commented on start_node_incremental_gcp.sh:
```diff
@@ -1,20 +1,23 @@
+KEYSTORE_DIR=testnet_dev_staging_keys
 if [[ "$1" = 'spring' ]]; then
     export TRACKER_WS_ADDR=ws://35.221.137.80:5000
+    KEYSTORE_DIR=testnet_spring_summer_keys
```
what about 'testnet_prod_keys'?

liayoo replied:
Done!

platfowner commented on deploy_blockchain_incremental_gcp.sh:
```diff
@@ -1,7 +1,7 @@
-if [[ "$#" -lt 5 ]] || [[ "$#" -gt 6 ]]; then
-    printf "Usage: bash deploy_blockchain_incremental_gcp.sh [dev|staging|spring|summer] <GCP Username> <# of Shards> [fast|full] [canary|full] [--setup]\n"
+if [[ "$#" -lt 5 ]] || [[ "$#" -gt 8 ]]; then
+    printf "Usage: bash deploy_blockchain_incremental_gcp.sh [dev|staging|spring|summer] <GCP Username> <# of Shards> [fast|full] [canary|full] [--setup] [--keystore <Password>]\n"
```
Instead of command line options, which can be exposed with 'ps -ef' command, what about using online text input functionalities of bash shell (see 'read -p ...' at line 77)?

liayoo replied:
Great idea, done!

liayoo commented:
Addressed all comments & updated the design as discussed offline. PTAL @platfowner @minsulee2 

Key changes
- Passing password through env var + fifo => getting password through read + passing encrypted password using an api
- Introduced bootstrapAccount in BlockchainNode
- Added /ain_getBootstrapPubKey, /ain_initAccount json-rpc apis for getting the bootstrapAccount's public key and initializing account from a keystore by sending an encrypted keystore file password

platfowner commented on init_account_gcp.js:
```diff
@@ -0,0 +1,72 @@
+      jsonrpc: '2.0',
+      id: 0
+    }
+  )
```
merging it to the previous line?

liayoo replied:
Done!

platfowner commented on deploy_blockchain_incremental_gcp.sh:
```diff
@@ -122,9 +155,24 @@ function deploy_node() {
+        printf "\n* >> Initializing account for node $node_index ********************\n\n"
+        printf "node_ip_addr='$node_ip_addr'\n"
+
+        node init_account_gcp.js $node_ip_addr $PASSWORD
```
Could you consider using unix pipe and standard input handling of nodejs for password? e.g.
`echo $PASSWORD | node init_account_gcp.js $node_ip_addr`

- https://nodejs.org/api/readline.html#readline_readline_createinterface_options

liayoo replied:
Ah, I missed this. Thanks! Done.

platfowner commented on client/index.js:
```diff
@@ -1,7 +1,7 @@
 
 const express = require('express');
-const jayson = require('jayson');
+const jayson = require('jayson/promise');
```
could you add a quick note on this change?

liayoo replied:
Done! It's been changed because ain_injectAccount needs to be an async function.

platfowner commented on common/file-util.js:
```diff
@@ -112,6 +113,12 @@ class FileUtil {
     }
   }
 
+  static createKeystoreDir(keysPath) {
```
suggestion: it seems this function can be generalized to createDir() and be used in other places in this file?

liayoo replied:
Nice! Done!

platfowner commented on common/file-util.js:
```diff
@@ -200,6 +207,11 @@ class FileUtil {
     }
   }
 
+  static getAccountAtKeystorePath(keystorePath, password) {
```
minor: what about getAccountFromKeystoreFile()?

liayoo replied:
Done!

platfowner commented on node/index.js:
```diff
@@ -37,32 +40,79 @@ const Transaction = require('../tx-pool/transaction');
+      const accountFromKeystore = FileUtil.getAccountAtKeystorePath(KEYSTORE_FILE_PATH, password);
+      if (accountFromKeystore !== null) {
+        this.account = accountFromKeystore;
+        this.bootstrapAccount = null;
```
minor: what about extracting the two lines above as a separate function e.g. setAccount()?

liayoo replied:
Done!

platfowner commented on node/index.js:
```diff
@@ -37,32 +40,79 @@ const Transaction = require('../tx-pool/transaction');
     logger.info(`Now node in STARTING state!`);
-    this.snapshotDir = path.resolve(SNAPSHOTS_ROOT_DIR, `${PORT}`);
-    FileUtil.createSnapshotDir(this.snapshotDir);
+    this.bootstrap();
```
some ideas: it sounds like 'bootstrapping' is a wide concept for the process of starting a blockchain node. so here are some ideas:
- bootstrap() -> initAccount() or something
- initAccount() -> injectAccount() or something

in this context further renaming can be considered:
- ain_initAccount() -> ain_injectAccount() in json rpc
- sendInitAccountRequest() -> sendInjectAccountRequest() in init_account_gcp.js
- initAccount() -> injectAccount() in init_account_gcp.js
- init_account_gcp.js -> inject_node_account_gcp.js
?

liayoo replied:
I like the new names! Updated, thanks!

platfowner commented on reset_blockchain_gcp.sh:
```diff
@@ -22,6 +24,49 @@ echo "PROJECT_ID=$PROJECT_ID"
+        printf "\n* >> Initializing account for node $node_index ********************\n\n"
+        printf "node_ip_addr='$node_ip_addr'\n"
+
+        node init_account_gcp.js $node_ip_addr $PASSWORD
```
ditto

liayoo replied:
Done

platfowner commented on deploy_blockchain_genesis_gcp.sh:
```diff
@@ -26,8 +27,33 @@ echo "GCP_USER=$GCP_USER"
 echo "NUM_SHARDS=$NUM_SHARDS"
 
-OPTIONS="$4"
+KEYSTORE_COMMAND_SUFFIX=""
```
minor: what about just KEYSTORE_OPTION and SETUP_OPTION?

liayoo replied:
done

platfowner commented on deploy_blockchain_genesis_gcp.sh:
```diff
@@ -26,8 +27,33 @@ echo "GCP_USER=$GCP_USER"
+    if [[ "$4" = '--setup' ]]; then
+        OPTIONS="$4"
+    elif [[ "$4" = '--keystore' ]]; then
+        KEYSTORE_COMMAND_SUFFIX="--keystore"
```
minor: KEYSTORE_COMMAND_SUFFIX="$4" ?

liayoo replied:
done

platfowner commented on deploy_blockchain_genesis_gcp.sh:
```diff
@@ -26,8 +27,33 @@ echo "GCP_USER=$GCP_USER"
+    fi
+elif [[ "$#" = 5 ]]; then
+    KEYSTORE_COMMAND_SUFFIX="--keystore"
+    if [[ "$4" = '--setup' ]] && [[ "$5" = '--keystore' ]]; then
```
suggestion: for simplicity what about extracting option setting parts as a separate function e.g. parse_option()?
(function argument passing ref: deploy_node() function)

liayoo replied:
great idea, done!

platfowner commented on deploy_blockchain_genesis_gcp.sh:
```diff
@@ -26,8 +27,33 @@ echo "GCP_USER=$GCP_USER"
+echo "KEYSTORE_COMMAND_SUFFIX=$KEYSTORE_COMMAND_SUFFIX"
+
+# Commands for starting nodes.
+BASE_COMMAND=". start_node_genesis_gcp.sh $SEASON"
```
minor: longer name like START_NODE_COMMAND_BASE?

platfowner replied:
or START_CMD_BASE?

liayoo replied:
done

platfowner commented on deploy_blockchain_genesis_gcp.sh:
```diff
@@ -39,8 +65,30 @@ then
+        printf "\n* >> Initializing account for node $node_index ********************\n\n"
+        printf "node_ip_addr='$node_ip_addr'\n"
+
+        node init_account_gcp.js $node_ip_addr $PASSWORD
```
ditto

liayoo replied:
done

platfowner commented on deploy_blockchain_incremental_gcp.sh:
```diff
@@ -53,8 +75,19 @@ if [[ ! $REPLY =~ ^[Yy]$ ]]; then
+    echo
+
+    # Read node ip addresses
+    IFS=$'\n' read -d '' -r -a IP_ADDR_LIST < ./testnet_ip_addresses/$SEASON
```
what about adding some file extension? "./testnet_ip_addresses/$SEASON.txt"

liayoo replied:
done

platfowner commented on reset_blockchain_gcp.sh:
```diff
@@ -63,20 +108,31 @@ if [[ "$3" -gt 0 ]]; then
+CHAINS_DIR=/home/ain_blockchain_data/chains
+SNAPSHOTS_DIR=/home/ain_blockchain_data/snapshots
+START_TRACKER_CMD="sudo rm -rf /home/ain_blockchain_data/ && cd \$(find /home/ain-blockchain* -maxdepth 0 -type d) && sudo rm -rf ./logs/ && . start_tracker_genesis_gcp.sh"
+START_NODE_CMD_PREFIX="sudo rm -rf $CHAINS_DIR $SNAPSHOTS_DIR && cd \$(find /home/ain-blockchain* -maxdepth 0 -type d) && sudo rm -rf ./logs/ && . start_node_genesis_gcp.sh $SEASON"
```
minor: what about START_NODE_CMD_BASE?

platfowner replied:
or START_CMD_BASE?

liayoo replied:
done

platfowner commented on deploy_blockchain_genesis_gcp.sh:
```diff
@@ -26,8 +27,30 @@ echo "GCP_USER=$GCP_USER"
+# Parse options.
+KEYSTORE_OPTION=""
+parse_options "$4"
+parse_options "$5"
```
param number checks?

liayoo replied:
done!

platfowner commented on deploy_blockchain_incremental_gcp.sh:
```diff
@@ -42,8 +42,26 @@ else
+# Parse options.
+KEYSTORE_OPTION=""
+parse_options "$6"
+parse_options "$7"
```
ditto

liayoo replied:
done!

platfowner commented on deploy_blockchain_incremental_gcp.sh:
```diff
@@ -122,9 +152,24 @@ function deploy_node() {
+        printf "\n* >> Initializing account for node $node_index ********************\n\n"
+        printf "node_ip_addr='$node_ip_addr'\n"
+
+        echo $PASSWORD | node inject_account_gcp.js $node_ip_addr $PASSWORD
```
could you remove $PASSWORD at the end?

liayoo replied:
oops, done!

platfowner commented on inject_account_gcp.js:
```diff
@@ -0,0 +1,81 @@
+const { sleep } = require('./common/common-util');
+const readline = require('readline');
+
+const rl = readline.createInterface({
```
could you use more longer variable name for rl?

liayoo replied:
done!

platfowner commented on inject_account_gcp.js:
```diff
@@ -0,0 +1,81 @@
+      },
+      jsonrpc: '2.0',
+      id: 0
+    })
```
minor: could you shift right to match the indentation below?

liayoo replied:
done!

platfowner commented on inject_account_gcp.js:
```diff
@@ -0,0 +1,81 @@
+      },
+      jsonrpc: '2.0',
+      id: 0
+    })
```
ditto

liayoo replied:
done!

platfowner commented on reset_blockchain_gcp.sh:
```diff
@@ -22,6 +24,49 @@ echo "PROJECT_ID=$PROJECT_ID"
 GCP_USER="$2"
 echo "GCP_USER=$GCP_USER"
 
+KEYSTORE_COMMAND_SUFFIX=""
```
what about KEYSTORE_OPTION?

liayoo replied:
done!

platfowner commented on start_node_genesis_gcp.sh:
```diff
@@ -1,35 +1,70 @@
     exit
 fi
 
+KEEP_CODE=""
```
what about applying parse_options function and using KEEP_CODE_OPTION and KEYSTORE_OPTION names?

liayoo replied:
updated to use parse_options here as well. thanks!

platfowner commented on start_node_incremental_gcp.sh:
```diff
@@ -81,22 +84,22 @@ fi
 
 export SYNC_MODE="$4"
 printf "SYNC_MODE=$SYNC_MODE\n"
+OPTIONS=$5
```
what about KEYSTORE_OPTION?

liayoo replied:
done!

platfowner commented on start_tracker_genesis_gcp.sh:
```diff
@@ -1,23 +1,46 @@
+    exit
+fi
+
+KEEP_CODE=""
```
what about KEEP_CODE_OPTION?

liayoo replied:
done!

platfowner commented on unittest/p2p.test.js:
```diff
@@ -124,34 +128,35 @@ describe("p2p", () => {
+        nodeStatus.dbStatus.stateInfo['#tree_bytes'] = 'erased';
+        nodeStatus.dbStatus.stateInfo['#state_ph'] = 'erased';
+        nodeStatus.dbStatus.stateProof['#state_ph'] = 'erased';
+        assert.deepEqual(actual, nodeStatus);
```
Expected use pattern of deepEqual is assert.deepEqual(actual, expected[, message]).
(see https://nodejs.org/api/assert.html#assert_assert_deepequal_actual_expected_message) 

In my previous I just tried to fix the order, could you take a look?

liayoo replied:
right, it was long overdue. switched the positions (and definitions) of actual & expected in p2p test cases. thanks!

platfowner replied:
Thanks for handling this! cc: @minsulee2

platfowner commented:
nice job done! thanks!

liayoo commented:
Thanks for the review!! Could you click approve as well? :) @platfowner

platfowner reviewed (approved):
LGTM
Thanks for the nice feature!

platfowner commented:
> Thanks for the review!! Could you click approve as well? :) @platfowner

Oops, just done!
